extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,600 @@
+    return;
+  }
+  if (!safeframeHost.channel) {
+    safeframeHost.connectMessagingChannel(data[MESSAGE_FIELDS.CHANNEL]);
+  } else if (payload) {
+    // Currently we do not expect a payload on initial connection messages.
+    safeframeHost.processMessage(/** @type {!JsonObject} */(payload),
+        data[MESSAGE_FIELDS.SERVICE]);
+  }
+}
+
+/**
+ * Sets up the host API for DoubleClick Safeframe to allow the following
+ * Safeframe container APIs to work:
+ *   - $sf.ext.expand()
+ *   - $sf.ext.collapse()
+ *   - $sf.ext.geom()
+ * Expand and collapse are both implemented utilizing AMP's built in element
+ * resizing.
+ *
+ * For geom, the host needs to send geometry updates into the container
+ *  whenever a position change happens, at a max frequency of 1 message/second.
+ *  To implement this messaging, we are leveraging the existing IntersectionObserver
+ *  class that works with AMP elements. However, the safeframe iframe that we
+ *  need to monitor is not an AMP element, but rather contained within an amp-ad.
+ *  So, we are doing intersection observing on the amp-ad, and calculating
+ *  the correct position for the iframe whenever we get an update.
+ *
+ * We pass an instance of this class into the IntersectionObserver class, which then
+ *  calls the instance of send() below whenever an update occurs.
+ * @implements {SimplePostMessageApiDef}
+ */
+export class SafeframeHostApi {
+
+  /**
+   * @param {!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl} baseInstance
+   * @param {boolean} isFluid
+   * @param {?({width: number, height: number}|../../../src/layout-rect.LayoutRectDef)} initialSize
+   * @param {?({width, height}|../../../src/layout-rect.LayoutRectDef)} creativeSize
+   * @param {?string} fluidImpressionUrl
+   */
+  constructor(baseInstance, isFluid, initialSize, creativeSize,
+    fluidImpressionUrl) {
+    /** @private {!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl} */
+    this.baseInstance_ = baseInstance;
+
+    /** @private {Window} */
+    this.win_ = this.baseInstance_.win;
+
+    /** @private {string} */
+    this.sentinel_ = this.baseInstance_.sentinel;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?IntersectionObserver} */
+    this.intersectionObserver_ = null;
+
+    /** @type {?string} */
+    this.channel = null;
+
+    /** @private {?JsonObject} */
+    this.currentGeometry_ = null;
+
+    /** @private {number} */
+    this.endpointIdentity_ = Math.random();
+
+    /** @private {number} */
+    this.uid_ = Math.random();
+
+    /** @private {boolean} */
+    this.isFluid_ = isFluid;
+
+    /** @private {?({width: number, height: number}|../../../src/layout-rect.LayoutRectDef)} */
+    this.initialSize_ = initialSize;
+
+    /** @private {?({width, height}|../../../src/layout-rect.LayoutRectDef)} */
+    this.creativeSize_ = creativeSize;
+
+    /** @private {!Object} */
+    this.iframeOffsets_ = {};
```

Comment:
Docs would be nice for iframeOffsets_.

Does this have to be a !Object, or is there an existing interface out there for it so that the compiler can verify correctness.

Reply:
removed, N/a